### ISSUE TITLE
Cannot use the default DECODER with a provider-side STORE and KEYMGMT

### ISSUE THREAD

**Issue by gotthardp** [openssl/openssl#15775]
I have a provider-side KEYMGMT and a provider-side STORE that calls default DECODER to decode a standard pubkey, which I want to IMPORT to my KEYMGMT. Decoding works great, but then `ossl_store_handle_load_result` / `try_key_ref` gets called, which compares the STORE provider with the KEYMGMT provider to determine whether to do LOAD or the EXPORT/IMPORT "dance". It's my STORE and my KEYMGMT so it does LOAD, but it's the default DECODER that created the structure, so I end with an error.
https://github.com/openssl/openssl/blob/29b3fdad2b078f45f840f6e45b0fe483b77dbc6f/crypto/store/store_result.c#L209

This is a likely a re-appearance of a problem discussed in https://github.com/openssl/openssl/issues/13539 that *the KEYMGMT provider is compared to the STORE provider, not the DECODER provider...  it should really do the latter*. See https://github.com/openssl/openssl/issues/13539#issuecomment-742426082 posted by @levitte.

Is this use-case supposed to work, please, or the STORE and the DECODER must belong to the same provider?

**Comment by romen** [comment]
@gotthardp we discussed this in our weekly OTC meeting, and the end result was that we need more details to understand if this use case is supported.

It would actually be very helpful if you could design a minimal working example (potentially to be incorporated in our tests if indeed we should support this use case for 3.0).

Would you be able to help us investigate this issue?

**Comment by mattcaswell** [comment]
> Is this use-case supposed to work, please, or the STORE and the DECODER must belong to the same provider?

A store loader capable of loading keys must have a key manager associated with it in the same provider. When loading a key the store loader's job is to load that key into a form that the associated key manager understands, and then the loader's "load" function invokes the callback its been provided and supplies a reference to the associated key available from its key manager.

It absolutely must not pass back a key reference that it doesn't understand, e.g. a key reference from some other provider's key manager. A key reference is only useable in the scope of the provider that loaded it. A store loader's key references implicitly are for the scope of the loaders own key manager.

It does not make sense to me for a store loader to pass back a key reference to some other provider's key manager. For third party provider's it is normal for them to be using a different libctx to the application (even a child libctx is counted as a "different" libctx). Therefore in the normal case the application may not even be able to get hold of the same key manager instance for this "other" provider because its been loaded into a libctx that the application does not have access to. It is a quirk of our own internal built-in providers (base and default) that they happen to use exactly the same libctx as the application - but this is not normally the case.

I suspect even our own file loader gets this wrong. I *think* it will just use any decoder (even those from another provider) and pass back unmodified "foreign" key references back to the application.

@levitte? Is my thinking above correct?

**Comment by t8m** [comment]
> It does not make sense to me for a store loader to pass back a key reference to some other provider's key manager.

That is a quite significant limitation. It basically requires a STORE provider to be a full support provider for everything, or do I misunderstand? I thought a STORE provider should be just a data manager wrapper for any decoders that are available within the context.

**Comment by t8m** [comment]
Hmm, but could you @gotthardp just pass the DER data of the pubkey by value? Then the `try_key_value()` would be used instead which should work.

**Comment by mattcaswell** [comment]
> That is a quite significant limitation

As you point out it is a restriction on passing keys by reference. The reference is implicitly within the scope of the store provider. If the decoder is "foreign" to the store then passing the key back by value should work.

Alternatively, if I understand correctly, @gotthardp does have a suitable key manager in the same provider as the store loader. So the loader can use a foreign decoder to decode the key. It then exports the key from the foreign decoder into the local key manager (using EVP_PKEY_export()), and finally passes back the key reference to the local instance of the key.

**Comment by gotthardp** [comment]
@t8m the `try_key_value()` seems to work, but it requires the user to load the default provider explicitly. In my view the user should not need to investigate what providers are needed to perform an operation-- the "dependent" providers (e.g. the default provider for decoding pubkeys) should(?) be loaded by the main provider, but the `try_key_value()` uses the "app-level" libctx, where I cannot "inject" any extra decoders that haven't been loaded by the user. (Or?)
BTW, before 29bf83c889c44236f33004ea2a6126c6d92e8b7a the decoding used to work even without loading the default provider explicitly. But I understand I misused some legacy path.

@mattcaswell yes, I have an (almost) full key manager and a store that calls a foreign decoder. The trouble is the key is not exported by the decoder, but passed directly as a reference, because the system thinks the decoder is from the same provider as the key management. (But it isn't.)

**Comment by levitte** [comment]
I'm a little confused...  what scheme do you provide with that STORE implementation, @gotthardp?

So here's the deal:

-   STORE in itself is meant to be fairly content agnostic, it depends on this like appropriate DECODERs to be present.
-   DECODERs can offer two ways to get the decoded object:
    -   by key reference: this comes with conditions
    -   by export via a callback and an OSSL_PARAM array: this requires that all key data is exportable

So, key references are fairly special, they are references to whatever a provider considers its internal key structure (for OpenSSL providers, these are the structure types, like our old types RSA, DSA, DH, EC_KEY, but for other providers, it may be whatever they like).

The thing with these internal key structures is that we must consider them as *internal to the provider*, i.e. they aren't transferable to other providers.  That's the reason why we have checks like the one you mentioned:

https://github.com/openssl/openssl/blob/29b3fdad2b078f45f840f6e45b0fe483b77dbc6f/crypto/store/store_result.c#L209

I think we've talked about this sort of thing before...  and it turned out that it was more sane to provide appropriate DECODERs for whatever you were trying to load, along with the appropriate KEYMGMTs and whatever algorithm implementations that come along with that.

Do you have your effort in a branch somewhere for us to look at?  That might make this discussion easier.

**Comment by t8m** [comment]
> In my view the user should not need to investigate what providers are needed to perform an operation-- the "dependent" providers (e.g. the default provider for decoding pubkeys) should(?) be loaded by the main provider, but the `try_key_value()` uses the "app-level" libctx, where I cannot "inject" any extra decoders that haven't been loaded by the user.

I think this is a misunderstanding of the concept. The applications are _usually_ supposed to have the default provider (or base & fips providers) loaded in their app libctx. Of course there might be specialized apps that load only one particular provider they are interested to work with but they have to accept the consequences. You can very well document in your provider documentation that it supports only this & that functionality and for general crypto operation of an application it has to have loaded also default provider (or base & fips providers).

Alternatively you could decode the data directly from the store provider implementation by calling the decoding process in the child library context, but then you still have to import that decoded key into your key manager and it can then be imported into your key manager by export&import of the key data. I would not recommend this path though. And it is still necessary for the app to have the default or base provider loaded in the app libctx, because the provider side child libctx just invokes implementations from the app libctx.

**Comment by levitte** [comment]
Is this related to tpm2-software/tpm2-openssl#5 ?

**Comment by levitte** [comment]
> So here's the deal:

Through discussion with team-mates, I found that I need to think on this another round

**Comment by levitte** [comment]
When playing with tpm2-openssl, I found one very curious thing:

``` console
$ LD_LIBRARY_PATH=$HOME/.private-install/opt/openssl-3.0/bin OPENSSL_MODULES=./.libs $HOME/.private-install/opt/openssl-3.0/bin/openssl list -provider tpm2 -decoders 
Provided DECODERs:
  DER @ tpm2 (provider=tpm2,input=pem)
```

And yet, the tpm2 provider has two decoders:

``` C
static const OSSL_ALGORITHM tpm2_decoders[] = {
    { "DER", "provider=tpm2,input=pem", tpm2_der_decoder_functions },
    { "RSA:rsaEncryption", "provider=tpm2,input=der,structure=TSS2", tpm2_tss_decoder_functions },
    { NULL, NULL, NULL }
};
```

Ref: https://github.com/tpm2-software/tpm2-openssl/blob/618b27c8e8dd595411a1a1c192a1360004b41138/src/tpm2-provider.c#L227-L231

Why is this happening?

**Comment by levitte** [comment]
Oooooohhhhh!

**Comment by levitte** [comment]
So, #15904 fixes part of the issues with tpm2-openssl.  I'm investigating further...

**Comment by levitte** [comment]
tpm2-openssl's 'file:' STORE has been thrown away a few days ago, when tpm2-software/tpm2-openssl#11 was merged, so this issue shouldn't be relevant any more...

**Comment by mattcaswell** [comment]
> tpm2-openssl's 'file:' STORE has been thrown away a few days ago, when tpm2-software/tpm2-openssl#11 was merged, so this issue shouldn't be relevant any more...

And #15981 will resolve any remaining issues from the OpenSSL side...?

**Comment by levitte** [comment]
> And #15981 will resolve any remaining issues from the OpenSSL side...?

I *think* so, but will use tpm2-openssl with possible changes there to verify

**Comment by levitte** [comment]
The issue presented here got resolved by #15981 and #15904, and the tpm2-openssl 'file:' store loader has been dropped.  I've seen files with TSS2 formatted keys, both PEM and DER, being properly loaded, so this issue is now getting closed.
